addons/l10n_dk_nemhandel/tools/demo_utils.py

```diff
@@ -0,0 +1,214 @@
+    self.ensure_one()
+    old_value = self.nemhandel_verification_state
+    if self.nemhandel_identifier_type and self.nemhandel_identifier_value:
+        self.nemhandel_verification_state = 'valid'
+
+def _mock_user_creation(func, self, *args, **kwargs):
+    func(self, *args, **kwargs)
+    self.l10n_dk_nemhandel_proxy_state = 'receiver' if self.receiver_registration else 'sender'
+    self.nemhandel_edi_user.write({
+        'private_key': b64encode(file_open(DEMO_PRIVATE_KEY, 'rb').read()),
+    })
+    return self._action_send_notification(
+        *_get_notification_message(self.l10n_dk_nemhandel_proxy_state)
+    )
+
+def _mock_receiver_registration(func, self, *args, **kwargs):
+    if not self.phone_number:
+        raise ValidationError(_("Please enter a phone number to verify your application."))
+    if not self.contact_email:
+        raise ValidationError(_("Please enter a primary contact email to verify your application."))
+    self.edi_user_id = self.edi_user_id.sudo()._register_proxy_user(self.company_id, 'nemhandel', self.edi_mode)
+    self.l10n_dk_nemhandel_proxy_state = 'receiver'
+    return self.env['nemhandel.registration']._action_send_notification(
+        *_get_notification_message(self.l10n_dk_nemhandel_proxy_state)
+    )
+
+def _mock_check_verification_code(func, self, *args, **kwargs):
+    self.button_nemhandel_sender_registration()
+    self.edi_user_id.nemhandel_verification_code = False
+    return self.env['nemhandel.registration']._action_send_notification(
+        *_get_notification_message(self.l10n_dk_nemhandel_proxy_state)
+    )
+
+def _mock_deregister_participant(func, self, *args, **kwargs):
+    # Set documents sent in demo to a state where they can be re-sent
+    demo_moves = self.env['account.move'].search([
+        ('company_id', '=', self.company_id.id),
+        ('nemhandel_message_uuid', '=like', 'demo_%'),
+    ])
+    demo_moves.write({
+        'nemhandel_message_uuid': None,
+        'nemhandel_move_state': None,
+    })
+    demo_moves.message_main_attachment_id.unlink()
+    demo_moves.ubl_cii_xml_id.unlink()
+    log_message = _('The Nemhandel status of the documents has been reset when switching from Demo to Live.')
+    demo_moves._message_log_batch(bodies=dict((move.id, log_message) for move in demo_moves))
+
+    # also unlink the demo vendor bill
+    self.env['account.move'].search([
+        ('company_id', '=', self.company_id.id),
+        ('nemhandel_message_uuid', '=', f'{self.company_id.id}_demo_vendor_bill'),
+    ]).unlink()
+
+    if 'nemhandel_edi_user' in self._fields:
+        self.nemhandel_edi_user.unlink()
+    else:
+        self.edi_user_id.unlink()
+    self.l10n_dk_nemhandel_proxy_state = 'not_registered'
+    if 'nemhandel_edi_mode' in self._fields:
+        self.nemhandel_edi_mode = 'demo'
+
+
+def _mock_update_user_data(func, self, *args, **kwargs):
+    pass
+
+
+def _mock_check_company_on_nemhandel(func, self, *args, **kwargs):
+    pass
```

Comment:
kept for the sake of futur overrides ?

Reply:
No, just to avoid doing a real query on the SML to check if the partner really exists on Nemhandel